connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java

```diff
@@ -0,0 +1,500 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.transforms.field;
+
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Schema.Type;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.transforms.util.SchemaUtil;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A FieldPath is composed by one or many field names, known as steps,
+ * to access values within a data object (either {@code Struct} or {@code Map<String, Object>}).a
+ *
+ * <p>If the SMT requires accessing multiple fields on the same data object,
+ * use {@link MultiFieldPaths} instead.
+ *
+ * <p>The field path semantics are defined by the {@link FieldSyntaxVersion syntax version}.
+ *
+ * <p>Paths are calculated once and cached for further access.
+ *
+ * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-821%3A+Connect+Transforms+support+for+nested+structures">KIP-821</a>
+ * @see FieldSyntaxVersion
+ * @see MultiFieldPaths
+ */
+public class SingleFieldPath {
+    // Invariants:
+    // - A field path can contain one or more steps
+    private static final char BACKTICK = '`';
+    private static final char DOT = '.';
+    private static final char BACKSLASH = '\\';
+
+    private final String[] path;
+
+    public SingleFieldPath(String pathText, FieldSyntaxVersion version) {
+        Objects.requireNonNull(pathText, "Field path cannot be null");
+        switch (version) {
+            case V1: // backward compatibility
+                this.path = new String[] {pathText};
+                break;
+            case V2:
+                this.path = buildFieldPathV2(pathText);
+                break;
+            default:
+                throw new IllegalArgumentException("Unknown syntax version: " + version);
+        }
+    }
+
+    private String[] buildFieldPathV2(String pathText) {
+        final List<String> steps = new ArrayList<>();
+        int idx = 0;
+        while (idx < pathText.length() && idx >= 0) {
+            if (pathText.charAt(idx) != BACKTICK) {
+                final int start = idx;
+                idx = pathText.indexOf(String.valueOf(DOT), idx);
+                if (idx > 0) { // get path step and move forward
```

Comment:
Sorry, I screwed this part up--should be `>=`, not `>`. It's possible (though unlikely) that someone could give us something like `...` and right now that would cause an infinite loop while trying to parse the path.

Also, to catch another similar edge case, we should add this check at the very end of the method, right before the return statement:

```java
if (!pathText.isEmpty() && pathText.charAt(pathText.length() - 1) == DOT)
    steps.add("");
```

I verified this with a new test case in the `SingleFieldPathTest` suite, in case it helps:
```java
@Test
public void testEmptyFieldNames() {
    assertArrayEquals(
            new String[] {"", "", ""},
            new SingleFieldPath("..", FieldSyntaxVersion.V2).path()
    );
    assertArrayEquals(
            new String[] {"foo", "", ""},
            new SingleFieldPath("foo..", FieldSyntaxVersion.V2).path()
    );
    assertArrayEquals(
            new String[] {"", "bar", ""},
            new SingleFieldPath(".bar.", FieldSyntaxVersion.V2).path()
    );
    assertArrayEquals(
            new String[] {"", "", "baz"},
            new SingleFieldPath("..baz", FieldSyntaxVersion.V2).path()
    );
}
```

Reply:
Sure, applying this and adding the test case.